internals/overlord/certstate/manager.go

```diff
@@ -0,0 +1,429 @@
+	if err != nil {
+		return err
+	}
+	keyPEM, err := privateKeyToPEM(keypair.Private)
+	if err != nil {
+		return err
+	}
+	if _, err = keyOut.Write(keyPEM); err != nil {
+		return err
+	}
+	if err := keyOut.Close(); err != nil {
+		return err
+	}
+	certPath := filepath.Join(tlsDir, fmt.Sprintf("%03d-cert.pem", keypair.Order))
+	certOut, err := os.OpenFile(certPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	certPEM, err := x509CertToPEM(keypair.Certificate)
+	if err != nil {
+		return err
+	}
+	if _, err = certOut.Write(certPEM); err != nil {
+		return err
+	}
+	if err := certOut.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// x509CertFromPEM loads an X509 certificate to a PEM encoded file.
+func x509CertFromPEM(path string) (*x509.Certificate, error) {
+	pemData, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemData)
+	if block == nil {
+		return nil, fmt.Errorf("empty PEM block")
+	}
+	return x509.ParseCertificate(block.Bytes)
+}
+
+// x509CertToPEM converts an X509 certificate to a PEM encoded byte slice.
+func x509CertToPEM(cert *x509.Certificate) ([]byte, error) {
+	pemBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert certificate to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// privateKeyFromPEM loads a private key from a PEM encoded file.
+func privateKeyFromPEM(path string) (any, error) {
+	pemData, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemData)
+	if block == nil {
+		return nil, fmt.Errorf("empty PEM block")
+	}
+	return x509.ParsePKCS8PrivateKey(block.Bytes)
+}
+
+// privateKeyToPEM converts a private key to a PEM encoded byte slice.
+func privateKeyToPEM(key any) ([]byte, error) {
+	privateBytes, err := x509.MarshalPKCS8PrivateKey(key)
+	if err != nil {
+		return nil, err
+	}
+	pemBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert key to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// isCertExpired checks if a X509 certificate is expired.
+func isCertExpired(cert *x509.Certificate) bool {
+	now := time.Now()
+	if now.Before(cert.NotBefore) {
+		return true
+	}
+	if now.After(cert.NotAfter) {
+		return true
+	}
+	return false
+}
+
+// expectPermission return an error if the specified directory or file
+// path is not matching the expected permission.
+func expectPermission(path string, perm fs.FileMode) error {
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	actualPerm := info.Mode().Perm()
+	if actualPerm != perm {
+		return fmt.Errorf("expected permission 0o%o (got 0o%o) for %q", perm, actualPerm, path)
+	}
+	return nil
+}
```

Comment:
This is great. However, I'd be more inclined to test group/other read/write bits against zero rather than expecting a concrete octal value here, and fail with something like `permissions <...> too open. Try 0X00`